projects/pgai/pgai/vectorizer/processor.py

```diff
@@ -266,4 +280,156 @@ async def run(self) -> Exception | None:
 
             poll_interval_str = datetime.timedelta(seconds=self.poll_interval)
             logger.info(f"sleeping for {poll_interval_str} before polling for new work")
-            await asyncio.sleep(self.poll_interval)
+
+            if shutdown_event is None:
+                await asyncio.sleep(self.poll_interval)
+            else:
+                # sleep for the poll interval or until the shutdown event is set
+                try:
+                    await asyncio.wait_for(
+                        shutdown_event.wait(), timeout=self.poll_interval
+                    )
+                    # shutdown event was set, the loop should exit
+                    logger.info("got a graceful shutdown request")
+                except asyncio.TimeoutError:
+                    # timeout means the sleep completed
+                    pass
+
+        if worker_tracking is not None:
+            await worker_tracking.force_last_heartbeat_and_stop()
+        logger.info("exiting processor.run()")
+        return None
+
+
+class ProcProcessor(Processor):
```

Comment:
Is there any reason why we subclass Processor? Could we not use composition here?